presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeFunctions.java

```diff
@@ -0,0 +1,160 @@
+    @ScalarFunction("second")
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BIGINT)
+    public static long second(@SqlType("time(p)") long time)
+    {
+        return (time / PICOSECONDS_PER_SECOND) % SECONDS_PER_MINUTE;
+    }
+
+    @Description("Minute of the hour of the given time")
+    @ScalarFunction("minute")
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BIGINT)
+    public static long minute(@SqlType("time(p)") long time)
+    {
+        return (time / PICOSECONDS_PER_MINUTE) % MINUTES_PER_HOUR;
+    }
+
+    @Description("Hour of the day of the given time")
+    @ScalarFunction("hour")
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BIGINT)
+    public static long hour(@SqlType("time(p)") long time)
+    {
+        return time / PICOSECONDS_PER_HOUR;
+    }
+
+    @Description("Truncate to the specified precision in the session timezone")
+    @ScalarFunction("date_trunc")
+    @LiteralParameters({"x", "p"})
+    @SqlType("time(p)")
+    public static long truncate(@SqlType("varchar(x)") Slice unit, @SqlType("time(p)") long time)
+    {
+        String unitString = unit.toStringUtf8().toLowerCase(ENGLISH);
+
+        switch (unitString) {
+            case "millisecond":
+                return time / PICOSECONDS_PER_MILLISECOND * PICOSECONDS_PER_MILLISECOND;
+            case "second":
+                return time / PICOSECONDS_PER_SECOND * PICOSECONDS_PER_SECOND;
+            case "minute":
+                return time / PICOSECONDS_PER_MINUTE * PICOSECONDS_PER_MINUTE;
+            case "hour":
+                return time / PICOSECONDS_PER_HOUR * PICOSECONDS_PER_HOUR;
+            default:
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "'" + unitString + "' is not a valid Time field");
+        }
+    }
+
+    @Description("Add the specified amount of time to the given time")
+    @LiteralParameters({"x", "p"})
+    @ScalarFunction("date_add")
+    @SqlType("time(p)")
+    public static long dateAdd(
+            @LiteralParameter("p") long precision,
+            @SqlType("varchar(x)") Slice unit,
+            @SqlType(StandardTypes.BIGINT) long value,
+            @SqlType("time(p)") long time)
+    {
+        long delta = value;
+        String unitString = unit.toStringUtf8().toLowerCase(ENGLISH);
+        switch (unitString) {
+            case "millisecond":
+                delta *= PICOSECONDS_PER_MILLISECOND;
+                break;
+            case "second":
+                delta *= PICOSECONDS_PER_SECOND;
+                break;
+            case "minute":
+                delta *= PICOSECONDS_PER_MINUTE;
+                break;
+            case "hour":
+                delta *= PICOSECONDS_PER_HOUR;
+                break;
+            default:
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "'" + unitString + "' is not a valid Time field");
+        }
+
+        long result = TimeOperators.add(time, delta);
```

Comment:
This will return negative result if  delta is "very negative", because `TimeOperators.add` has 
```
if (result < 0) {
```

and not *e.g.*
```
while (result < 0) {
```

Reply:
`add` performs math modulo 24h